compiler/qsc_passes/src/capabilitiesck.rs

```diff
@@ -0,0 +1,555 @@
+            FunctorSetValue::Adj => callable_compute_properties
+                .adj
+                .as_ref()
+                .expect("adj specialization is none"),
+            FunctorSetValue::Ctl => callable_compute_properties
+                .ctl
+                .as_ref()
+                .expect("ctl specialization is none"),
+            FunctorSetValue::CtlAdj => callable_compute_properties
+                .ctl_adj
+                .as_ref()
+                .expect("ctl_adj specialization is none"),
+        };
+
+        if let ComputeKind::Quantum(quantum_properties) = spec_compute_properties.inherent {
+            let missing_features = get_missing_runtime_features(
+                quantum_properties.runtime_features,
+                self.target_capabilities,
+            );
+            let missing_spec_level_runtime_features =
+                get_spec_level_runtime_features(missing_features);
+
+            // If there are any missing specialization-level runtime features, runtime features that affect the whole
+            // specialization, just generate errors for the missing specialization-level runtime features and do not
+            // generate statement-level or expression-level errors for these specializations.
+            if !missing_spec_level_runtime_features.is_empty() {
+                let current_callable = self
+                    .package
+                    .get_global(current_callable_id)
+                    .expect("callable not present in package");
+                let Global::Callable(callable_decl) = current_callable else {
+                    panic!("");
+                };
+
+                self.missing_features_map
+                    .entry(callable_decl.name.span)
+                    .and_modify(|f| *f |= missing_spec_level_runtime_features)
+                    .or_insert(missing_spec_level_runtime_features);
+                return;
+            }
+        }
+
+        // Visit the specialization block.
+        if let Some(spec_decl) = spec_decl {
+            self.visit_block(spec_decl.block);
+        }
+    }
+
+    fn clear_current_callable(&mut self) -> LocalItemId {
+        self.current_callable
+            .take()
+            .expect("current callable is not set")
+    }
+
+    fn find_local_var_ident(&self, local_var_id: LocalVarId) -> Option<&Ident> {
+        self.package
+            .pats
+            .iter()
+            .map(|(_, pat)| pat)
+            .find_map(|pat| {
+                if let PatKind::Bind(ident) = &pat.kind {
+                    if ident.id == local_var_id {
+                        Some(ident)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+    }
+
+    fn generate_errors(&mut self) -> Vec<Error> {
+        let mut errors = Vec::new();
+
+        for (span, missing_features) in self.missing_features_map.drain().sorted() {
```

Comment:
We should avoid `sorted` as it sneakily introduces a copy (not always, but often). I'd prefer to be explicit about it so ownership is easy to track:

```suggestion
        let mut missing_features_map = self.missing_features_map.drain().collect::<Vec<_>>();
        missing_features_map.sort_unstable();
        for (span, missing_features) in missing_features_map {
```

Reply:
Updated to not use `sorted`.